service-registry/registry-consul/src/main/java/org/apache/servicecomb/registry/consul/ConsulRegistration.java

```diff
@@ -0,0 +1,176 @@
+    consulInstance.addProperty(key, value);
+  }
+
+  @Override
+  public boolean enabled() {
+    return consulDiscoveryProperties.isEnabled();
+  }
+
+  @Override
+  public void init() {
+    logger.info("ConsulRegistration init");
+    String serverPort;
+    if (restAddress.contains("?")) {
+      serverPort = restAddress.substring(restAddress.indexOf(":") + 1, restAddress.indexOf("?"));
+    } else {
+      serverPort = restAddress.substring(restAddress.indexOf(":") + 1);
+    }
+    String env = BootStrapProperties.readServiceEnvironment(environment);
+    if (StringUtils.isEmpty(env)) {
+      env = ConsulConst.CONSUL_DEFAULT_ENVIRONMENT;
+    }
+    String instanceId = registrationId.getInstanceId();
+    String serviceName = BootStrapProperties.readServiceName(environment);
+    consulInstance = new ConsulInstance();
+    consulInstance.setServiceId(serviceName + "-" + serverPort);
+    consulInstance.setInstanceId(instanceId);
+    consulInstance.setEnvironment(env);
+    consulInstance.setApplication(BootStrapProperties.readApplication(environment));
+    consulInstance.setServiceName(BootStrapProperties.readServiceName(environment));
+    consulInstance.setAlias(BootStrapProperties.readServiceAlias(environment));
+    consulInstance.setDescription(BootStrapProperties.readServiceDescription(environment));
+    if (StringUtils.isNotEmpty(dataCenterProperties.getName())) {
+      DataCenterInfo dataCenterInfo = new DataCenterInfo();
+      dataCenterInfo.setName(dataCenterProperties.getName());
+      dataCenterInfo.setRegion(dataCenterProperties.getRegion());
+      dataCenterInfo.setAvailableZone(dataCenterProperties.getAvailableZone());
+      consulInstance.setDataCenterInfo(dataCenterInfo);
+    }
+    consulInstance.setProperties(BootStrapProperties.readServiceProperties(environment));
+    consulInstance.setVersion(BootStrapProperties.readServiceVersion(environment));
+
+    List<String> tags = new ArrayList<>();
+    tags.add(consulInstance.getServiceName());
+    registrationBuilder = ImmutableRegistration.builder()
+        .id(consulInstance.getServiceId())
+        .name(consulInstance.getServiceName())
+        .address(consulDiscoveryProperties.getIpAddress())
+        .port(Integer.parseInt(serverPort))
+        .tags(tags);
+  }
+
+  @Override
+  public void run() {
+    logger.info("ConsulRegistration run");
+    Map<String, String> meta = new HashMap<>();
+    meta.put("serviceName", consulInstance.getServiceName());
+    meta.put("version", consulInstance.getVersion());
+    meta.put("instanceId", registrationId.getInstanceId());
+    meta.put("env", consulInstance.getEnvironment());
+    meta.put("application", consulInstance.getApplication());
+    meta.put("alias", consulInstance.getAlias() != null ? consulInstance.getAlias() : "");
+    Gson gson = new GsonBuilder().disableHtmlEscaping().create();
+    meta.put("endpoints", gson.toJson(consulInstance.getEndpoints()));
```

Comment:
Maybe you can add `consulInstance` as a json propertly, and code can be simpler. See Zookeeper for an example.